Implemented tough-kms to include AWS KMS as a new keysource

srgothi92 opened the pull request:
*Issue #, if available:*
N/A

*Description of changes:*
-tough-kms implements the "KeySource" trait found in tough
-Unit test cases are added for tough-kms utilizing rusoto_mock
-Integration Tests are added in tuftool, to test 3 keysource: tough-kms, tough-ssm, local key.
-To run integration tests use feature flag "integ" while running test


Testing Done:
Ran Unit tests
Ran Integration Tests
Manually created root.json and used kms key for signing.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Commit:
Implemented tough-kms to include AWS KMS as a new keysource

* tough-kms implements the "KeySource" trait found in tough
* A new method "create" is added to "KeySource" trait to allow KeySource
  implementer to create Key.
* Integration Tests are added in tuftool create to test 3 keysource.
* To run integration tests use feature flag "integration-tests" while running test

PR comments:
* Removed toug-kms create and created issue #211 to handle it.
* KMS CMK needs to be manually created separately
* Moved Sign implementation for KmsRsaKey to tough_kms crate

zmrow commented on tough-kms/Cargo.toml:
```diff
@@ -0,0 +1,31 @@
+keywords = ["TUF", "KMS"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
We can get rid of this line. :)

zmrow commented on tough-kms/src/error.rs:
```diff
@@ -0,0 +1,121 @@
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display("Unknown AWS region \"{}\": {}", region, source))]
```
nit: you can use single quotes around the region (sort of the like the error variant above this).

`"Unknown AWS region '{}': {}"`

webern replied:
> nit: you can use single quotes around the region (sort of the like the error variant above this).

I agree. I noticed, because I'm comparing with tough-ssm, that it should be changed in that crate also.

zmrow commented on tough-kms/src/error.rs:
```diff
@@ -0,0 +1,121 @@
+        key_id: String,
+    },
+
+    #[snafu(display("Public key is none"))]
```
Could we make this error message a bit friendlier? Are we saying the public key doesn't exist? Or it exists but is empty? Or...?

zmrow commented on tough-kms/src/error.rs:
```diff
@@ -0,0 +1,121 @@
+    },
+
+    #[snafu(display(
+    "Alias {} exist for aws-kms:://{}/{}",
```
Is this supposed to say "Alias {} **exists**"?

zmrow commented on tough-kms/src/error.rs:
```diff
@@ -0,0 +1,121 @@
+    PublicKeyNone,
+
+    #[snafu(display(
+    "Write in Not defined for aws-kms://{}/{}",
```
I'm not sure what this means; could we clarify? :)

zmrow commented on tough-kms/src/error.rs:
```diff
@@ -0,0 +1,121 @@
+    },
+
+    #[snafu(display(
+    "Key metatada is None while creating key for aws-kms://{}/{}",
```
Rather than referring to Rust concepts of `Some() / None` in these error messages, let's describe the problem in prose. Are we trying to inform the user that the metadata is empty? doesn't exist?

zmrow commented on tough-ssm/src/lib.rs:
```diff
@@ -55,12 +55,11 @@ impl KeySource for SsmKeySource {
         let fut = ssm_client.put_parameter(rusoto_ssm::PutParameterRequest {
             name: self.parameter_name.to_owned(),
-            description: Some(key_id_hex.to_owned()),
+            description: Some(String::from("Tough Key")),
```
What's the reason for this change?

srgothi92 replied:
KeyId added in the root.json file is no longer available with write method because key is created by 'create' method which call the write method. After all this succeeds gen-rsa-key adds the key to root.json.

webern replied:
> KeyId added in the root.json file is no longer available with write method because key is created by 'create' method which call the write method. After all this succeeds gen-rsa-key adds the key to root.json.

This seems to be the only change in the tough-ssm crate. How did changes elsewhere affect the behaviour in this crate?

edit: later I saw the changes in `tough`...

zmrow commented on tough/src/key_source.rs:
```diff
@@ -17,11 +17,38 @@ pub trait KeySource: Debug + Send + Sync {
+    fn write(&self, value: &str) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>>;
+    /// Creates a key for key source, default implementation creates rsa key pair which can
+    /// be override by individual `KeySource` implementation
+    fn create(
```
Hmmm, I'm a little worried about a "default" implementation for this method that assumes `openssl` is installed.  

I'm also wondering why we need a `create` method?  Is it a convenience just for `tuftool`?

srgothi92 replied:
I'm also wondering why we need a create method? Is it a convenience just for tuftool?
Indirectly Yes, create is used by gen-rsa-key which is for convenience.
But, it helps us with end to end test case and programmatic key generation wherever required.

zmrow commented on tough/src/key_source.rs:
```diff
@@ -17,11 +17,38 @@ pub trait KeySource: Debug + Send + Sync {
 
     /// Writes a key back to the `KeySource`
-    fn write(
+    fn write(&self, value: &str) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>>;
```
I think the only reason we have a `write` method on this trait is to account for `tuftool`'s usage of it for SSM, right?  Perhaps it makes sense to evaluate what it means to eliminate this method from the trait altogether.  

@iliana : thoughts?

srgothi92 replied:
write is used by SSM to store the rsa key in a parameter.
write is used by KMS to assign alias to the generated Key.
write is used by localsource to write generated key in a file.


Previously, gen-rsa-key command creates the rsa-key pair and calls write method.
Now, gen-rsa-key calls the create method of keysource, which in turn calls the write method. After this code change, write method is not longer called directly. We can get rid of it or make it private.  

gen-rsa-key command is for convenience, but it helps in end to end testing and programmatically create key.

I was also wondering does it make sense to call 'add-key' command from 'gen-rsa-key' ? We have separate command 'add-key', which user can use if they wan't to add the key. Or we can all command 'gen-add-rsa-key'. :P

webern replied:
> gen-rsa-key command is for convenience, but it helps in end to end testing and programmatically create key.

I think we should take a closer look at this design, there may be another way. Generally speaking we shouldn't be adding too many dependencies or assumptions to the core library (`tough`).

zmrow commented on tuftool/src/root.rs:
```diff
@@ -242,34 +241,11 @@ impl Command {
-        let key_pair = parse_keypair(stdout.as_bytes()).context(error::KeyPairParse)?;
-        let key_id = hex::encode(add_key(&mut root.signed, roles, key_pair.tuf_key())?);
         key_source
-            .write(&stdout, &key_id)
```
I think this may be the only place we do a `key_source.write()` correct?  If so, let's look harder at eliminating this method from `key_source` as mentioned.

srgothi92 replied:
Yes, this is the only place its used.

zmrow commented on tuftool/src/source.rs:
```diff
@@ -55,7 +56,8 @@ pub(crate) fn parse_key_source(input: &str) -> Result<Box<dyn KeySource>> {
         .parse(input)
         .context(error::UrlParse { url: input })?;
-
+    println!("{}", url.scheme());
```
Are these `println!()`'s  here on purpose? Or were they leftover from debugging?  :)

zmrow commented on tuftool/tests/create_repository_integration.rs:
```diff
@@ -0,0 +1,248 @@
+
+// This file include integration tests for KeySources: tough-ssm, tough-kms and local file key.
+// Since the tests are run using the actual "AWS Ssm and AWS Kms", you would have to configure
+// AWS account profile "thar-signing-root-pdx" with root permission.
```
Could we please remove `pdx` and `thar` from the profile name? :)

zmrow commented on tuftool/tests/create_repository_integration.rs:
```diff
@@ -0,0 +1,248 @@
+// Since the tests are run using the actual "AWS Ssm and AWS Kms", you would have to configure
+// AWS account profile "thar-signing-root-pdx" with root permission.
+// Refer https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html to configure named profile.
+// To run test include feature flag 'integration-tests' like : "cargo test --features=integration-tests"
```
Thanks for putting this under a feature flag! 🎉

zmrow commented on tuftool/tests/create_repository_integration.rs:
```diff
@@ -0,0 +1,248 @@
+// Refer https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html to configure named profile.
+// To run test include feature flag 'integration-tests' like : "cargo test --features=integration-tests"
+
+fn initialie_root_json(root_json: &str) {
```
I think you meant `initialize...` :)

zmrow commented on tough/src/error.rs:
```diff
@@ -478,6 +478,46 @@ pub enum Error {
     #[snafu(display("Role missing from snapshot meta: {}", name))]
     RoleNotInMeta { name: String },
+
+    #[snafu(display("Not able to Sign message for key-id {} : {}", key_id, source))]
```
No need to capitalize `Sign`.

Nit: single quotes around the key-id: `...for key-id '{}':...`

zmrow commented on tough/src/error.rs:
```diff
@@ -478,6 +478,46 @@ pub enum Error {
+    #[snafu(display("Key Id: {} is not sign and verify key", key_id))]
+    NotSignVerifyKeyType { key_id: String },
+
+    #[snafu(display("Failed to run {}: {}", command_str, source))]
```
Could we single quote the command string to make it easier to see it's a command?

`Failed to run '{}':...`

zmrow commented on tough-kms/src/lib.rs:
```diff
@@ -0,0 +1,139 @@
+        &self,
+    ) -> std::result::Result<Box<dyn Sign>, Box<dyn std::error::Error + Send + Sync + 'static>>
+    {
+        let kms_client = match self.client.clone() {
```
Why do we have to `.clone()` here? I think we should just be able to match on `self.client`, no?

webern replied:
Maybe this, unless ownership is actually required.

```suggestion
        let kms_client = &match self.client {
```

zmrow reviewed (commented):
Nice work!

I didn't get through `tough-kms/tests/all_test.rs`, but here's what I have so far.

srgothi92 reviewed (commented):
Added comments for Design related questions.

srgothi92 reviewed (commented):
I added few other comments in last conversation. But, I could not see them any more. Adding them back.

webern commented on Cargo.toml:
```diff
@@ -3,5 +3,6 @@ members = [
     "olpc-cjson",
     "tough",
     "tough-ssm",
+    "tough-kms",
```
Also adda line to this file so it will be built during continuous integration. https://github.com/awslabs/tough/blob/498fb581bac6af4e3f93583e572939ca9bbbcb03/.github/workflows/rust.yml#L15

webern commented on tough-kms/src/client.rs:
```diff
@@ -0,0 +1,33 @@
+use snafu::ResultExt;
+use std::str::FromStr;
+
+/// Builds an KMS client for a given profile name.
```
```suggestion
/// Builds a KMS client for a given profile name.
```

webern commented on tough-kms/src/lib.rs:
```diff
@@ -0,0 +1,139 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT OR Apache-2.0
```
We should add these (or some subset), which I copied from tough/lib.rs. These will enforce certain lints. The most important ones are `#![deny(rust_2018_idioms)]` and `#![deny(missing_docs)]`. Once you add these you will have compilation errors that you have to fix.

```suggestion
// SPDX-License-Identifier: MIT OR Apache-2.0

#![forbid(missing_debug_implementations, missing_copy_implementations)]
#![deny(rust_2018_idioms)]
// missing_docs is on its own line to make it easy to comment out when making changes.
#![deny(missing_docs)]
#![warn(clippy::pedantic)]
#![allow(
    clippy::module_name_repetitions,
    clippy::must_use_candidate,
    clippy::missing_errors_doc
)]
```

webern commented on tough-kms/src/lib.rs:
```diff
@@ -0,0 +1,139 @@
+        &self,
+        value: &str,
+    ) -> std::result::Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+        let kms_client = match self.client.clone() {
```
See if this works (and self.pofile probably won't need as_deref() anymore?)

```suggestion
        let kms_client = match &self.client {
```

webern commented on tough-kms/src/lib.rs:
```diff
@@ -0,0 +1,139 @@
+            alias_name: self.key_id.clone(),
+            target_key_id: value.to_string(),
+        });
+        let _response = tokio::runtime::Runtime::new()
```
Double check to make sure there are no hidden, error-like, conditions that would require checking something in the response.

```text
        let _response = tokio::runtime::Runtime::new()
```

webern commented on tough-kms/src/lib.rs:
```diff
@@ -0,0 +1,139 @@
+            None => client::build_client_kms(self.profile.as_deref())?,
+        };
+        // Kms currently does not have an API to validate if the alias exist. As a Hack,
+        // or temporary resolution, we are using the get_public_key API, if it succeeds than
+        // Alias exist and alarm the user.
```
```suggestion
        // or temporary resolution, we are using the get_public_key API, if it succeeds then
        // an alias exists and we return an error.
```

webern commented on tough-kms/src/lib.rs:
```diff
@@ -0,0 +1,139 @@
+            key_usage: Some(String::from("SIGN_VERIFY")),
+            ..rusoto_kms::CreateKeyRequest::default()
+        });
+        let key_id = tokio::runtime::Runtime::new()
```
I think you could do this whole function in a single tokio runtime.

```
        let key_id = tokio::runtime::Runtime::new()
```

webern commented on tough/Cargo.toml:
```diff
@@ -17,10 +18,13 @@ olpc-cjson = { version = "0.1.0", path = "../olpc-cjson" }
 pem = "0.8.1"
 reqwest = { version = "0.10.4", optional = true, default-features = false, features = ["blocking"] }
 ring = { version = "0.16.13", features = ["std"] }
+rusoto_core = { version = "0.44", optional = true, default-features = false }
+rusoto_kms = { version = "0.44", optional = true, default-features = false }
```
The purpose of placing the KMS feature in its own crate is to avoid introducing these dependencies to `tough`.

```suggestion
```

webern commented on tough/Cargo.toml:
```diff
@@ -17,10 +18,13 @@ olpc-cjson = { version = "0.1.0", path = "../olpc-cjson" }
 serde_json = "1.0.53"
 serde_plain = "0.3.0"
 snafu = "0.6.8"
+tokio = "0.2.13"
```
This should also be a transient dependency, not needed directly by `tough`. (i.e. needed by tough-ssm, but not in tough's cargo.toml)

```suggestion
```

webern commented on tough/src/key_source.rs:
```diff
@@ -17,11 +17,38 @@ pub trait KeySource: Debug + Send + Sync {
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+        // ring doesn't support RSA key generation yet
+        // https://github.com/briansmith/ring/issues/219
+        let mut command = std::process::Command::new("openssl");
+        command.args(&["genpkey", "-algorithm", "RSA", "-pkeyopt"]);
+        command.arg(format!("rsa_keygen_bits:{}", bits));
+        command.arg("-pkeyopt");
+        command.arg(format!("rsa_keygen_pubexp:{}", exponent));
```
This we definitely should not do.
```
        let mut command = std::process::Command::new("openssl");
        command.args(&["genpkey", "-algorithm", "RSA", "-pkeyopt"]);
        command.arg(format!("rsa_keygen_bits:{}", bits));
        command.arg("-pkeyopt");
        command.arg(format!("rsa_keygen_pubexp:{}", exponent));
```

webern reviewed (commented):
I didn't get all the way through this. It seems like something is wrong with our design if adding a new key source causes a lot of churn in `tough`. The idea behind the key source trait was to ensure that `tough` (to the extent possible) doesn't know anything about what's under the hood in a key source.

Leaving the 'heavy lifting' of creating a key outside the realm of `tough` would be better if it prevents us from adding dependencies to `tough`.

I know you're working with Zack on this. I'd be happy to join a call to join any discussion. Thanks!

Nice start!

iliana commented:
First, run `cargo check --all` to regenerate the Cargo.lock, and commit that. CI is requiring that your Cargo.lock is accurate and maps to the dependencies you are using across the project.

No Rusoto dependencies should exist in the main tough crate. I moved all of the added structs to `tough-kms` locally and realized the reason you may have done this: the `Sign` trait's `sign` method returns a `Result<T, tough::Error>` instead of allowing the implementation to return a more generic error. Because we have to implement `Sign` differently for KMS, this is impossible without adding the code directly to `tough`.

If you're comfortable with it, modify the `Sign::sign` trait signature to return a generic error, similar to the methods in `KeySource` (they return a `Result<T, E>` where E is `Box<dyn std::error::Error + Send + Sync + 'static>>`). Alternately, if you like, I can make this change in a separate PR, and you can rebase on top of it and carry on.

srgothi92 commented:
Response to @iliana comment:
1:
Thanks.

2: 
Not just because of `Sign` trait's `sign` method returns a `Result<T, tough::Error>`. In Sign object I  would need KmsClient to sign the messages. For `LocalKeySource` and `tough-ssm`, sign object has private key available to sign the messages. Which is not the case with Kms. The solution I currently implemented adds `rusoto_kms` as dependency and makes a call to KMS directly using `Kmsclient` available with `Sign` trait. Which definitely is not the right way. 
Solution Suggestion 1:  I am thinking, if we can expose a module in `tough_kms` that is used by `Sign` trait to sign the messages, avoiding direct dependency on `KmsClient`, not sure if it would result in cyclic dependency. tough_kms provides sign object and sign object needs tough_kms. 
Solution Suggestion 2: We can get rid of `Sign` trait and let `KeySource` expose `sign` and `tuf_key` method. Its upto keySource implementer to cache the private key and use the cached key for multiple `sign` request.   

3: 
I can take care of it, if required.

iliana commented:
> In Sign object I would need KmsClient to sign the messages.

`KeySource::as_sign` returns a `Box<dyn Sign>`, and `KmsRsaKey` (which has your `KmsClient`) implements `Sign`. This separation allows you to move `KmsRsaKey` to tough-kms.

zmrow commented:
We just realized that we probably shouldn't create a tokio runtime inside these libraries, so we probably shouldn't ship this until https://github.com/awslabs/tough/issues/213 is fixed.  :(

webern reviewed (approved):
Nice! As I understand it, because tough-kms needs to implement sign, the sign trait was changed to return a generic error to break dependency cycle. Looks good, looks clean.

iliana commented:
Merging this into develop fails to compile — there are two new `Sign` implementations that need to have their signatures fixed. You'll probably want to rebase and resolve those issues.

I'm fixing that locally and continuing to review.

iliana commented on tuftool/tests/create_repository_integration.rs:
```diff
@@ -0,0 +1,268 @@
+}
+
+#[test]
+#[cfg_attr(not(feature = "integration-tests"), ignore)]
```
We use `integ` as the feature name in tough, so we should do the same here.

srgothi92 replied:
I din't use `integ` because it says `Integration tests require docker` in [docs](https://docs.rs/tough/0.8.0/tough/), but my itegration test needs a different setup. Let me know if u still need me to change it to `integ`.

iliana replied:
Please do, yeah. We can fix the docs.

iliana replied:
Well, more specifically, that's the `tough` docs, and `tuftool`'s definition of integration tests can mean "needs to have working AWS credentials". :)

srgothi92 replied:
Done

webern replied:
Yeah, it might be easiest (maybe per crate), to have a single `integ` feature and list all the things you need to run integration tests.

Edit: aaaactually, using an AWS Account is significantly more arduous than having docker available. But the current setup is fine since the two crates have different requirements for the `integ` feature, and these are documented...

iliana commented on tuftool/tests/create_repository_integration.rs:
```diff
@@ -0,0 +1,268 @@
+#[cfg_attr(not(feature = "integration-tests"), ignore)]
+// Ensure we can use ssm key to create and sign a repo created by the `tuftool` binary using the `tough` library
+fn create_repository_ssm_key() {
+    let root_key = &format!("aws-ssm://{}/thar/root-key-new", PROFILE);
```
These should be key sources a tough developer can set up once and continue to use, so we should have names that indicate they're for testing.

```suggestion
    let root_key = &format!("aws-ssm://{}/tough-integ/key-a", PROFILE);
```

srgothi92 replied:
Done :)

iliana commented on tuftool/tests/create_repository_integration.rs:
```diff
@@ -0,0 +1,268 @@
+#[cfg_attr(not(feature = "integration-tests"), ignore)]
+// Ensure we can use kms key to create and sign a repo created by the `tuftool` binary using the `tough` library
+fn create_repository_kms_key() {
+    let root_key = &format!("aws-kms://{}/alias/test_key", PROFILE);
```
```suggestion
    let root_key = &format!("aws-kms://{}/alias/tough-integ/key-a", PROFILE);
```

srgothi92 replied:
Done :)

iliana commented on tuftool/tests/create_repository_integration.rs:
```diff
@@ -0,0 +1,268 @@
+// Ensure we can use local rsa key to create and sign a repo created by the `tuftool` binary using the `tough` library
+fn create_repository_local_key() {
+    let root_key_dir = TempDir::new().unwrap();
+    let root_key_path = root_key_dir.path().join("local_key.pem");
```
This could instead use `snakeoil.pem` from tough's test data, and then this wouldn't need to be an integration test.

srgothi92 replied:
I am creating a new one because I wanted to test end to end. I created a separate method for whole repository creation process, So that we can just use different types of key and ensure they all adhere to the same workflow of consumption. Unfortunately, `tough-kms` had to add an extra flag for creation, but it will be same in future and this test cases ensures that.

iliana replied:
I misread the test, I understand now.

iliana commented on tough-kms/Cargo.toml:
```diff
@@ -0,0 +1,30 @@
+rusoto_kms = { version = "0.44", optional = true, default-features = false }
+bytes = "0.5.6"
+serde = "1.0.105"
+serde_json = "1.0.50"
```
These two dependencies only seem to be used in tests, so they should be `[dev-dependencies]`.

srgothi92 replied:
Yes, :). Awesome that u were able to see that from code review.

iliana commented on tuftool/tests/create_repository_integration.rs:
```diff
@@ -0,0 +1,268 @@
+        .success();
+    Command::cargo_bin("tuftool")
+        .unwrap()
+        .args(&["root", "expire", root_json.clone(), "2020-09-22T00:00:00Z"])
```
Seems like this will make tests start failing in a month... :)

```suggestion
        .args(&["root", "expire", root_json.clone(), "3030-09-22T00:00:00Z"])
```

srgothi92 replied:
Oops. Done.

iliana commented on tuftool/tests/create_repository_integration.rs:
```diff
@@ -0,0 +1,268 @@
+fn initialize_root_json(root_json: &str) {
+    Command::cargo_bin("tuftool")
+        .unwrap()
+        .args(&["root", "init", root_json.clone()])
```
Is cloning `root_json` necessary here? (`root_json` is already a reference so I think clone does nothing...)

srgothi92 replied:
Nice, Done. Even changed for `key.clone()`.

srgothi92 commented:
> Merging this into develop fails to compile — there are two new Sign implementations that need to have their signatures fixed. You'll probably want to rebase and resolve those issues.

> I'm fixing that locally and continuing to review.

Done

iliana commented on tough-kms/src/lib.rs:
```diff
@@ -0,0 +1,171 @@
+use tough::schema::decoded::{Decoded, RsaPem};
+use tough::schema::key::{Key, RsaKey, RsaScheme};
+use tough::sign::Sign;
+
+/// Represents a Signing Algorithms for AWS KMS.
+#[derive(Debug, Clone, PartialEq)]
+pub enum KmsSigningAlgorithms {
+    /// The key type
+    Rsa(String),
+}
```
I'm confused about why we need this type. What advantage does this have over `KmsRsaKey.signing_algorithm` just being a `String`?

srgothi92 replied:
Yes, It is confusing, I would like to have your feedback on this.  We had a discussion on question as described below
>  KMS supports a lot of signing methods

> > Iliana Reply: What I'd recommend is a match statement, even though the enum has one field. That way if we add new ones because TUF allows new ones, it's more clear that we need to change downstream code.

It was added in context of this, to keep it generic and make decision of signing algorithm in  `as_sign` while returning the `sign` object:
```
Ok(Box::new(KmsRsaKey {
            profile: self.profile.clone(),
            client: Some(kms_client.clone()),
            key_id: self.key_id.clone(),
            public_key: pb_key,
            signing_algorithm: KmsSigningAlgorithms::Rsa(String::from("RSASSA_PSS_SHA_256")),
        }))
```

However, It does not solve the problem in a way I wanted to.  I want something like object of algorithms to select from.
Ex:-
```
algorithms = {
   algo1 : "algo1",
   algo2: "algo2"
} 
```
 And we can directly use as `algorithms.algo1` or `algorithms.algo2` based on some conditions in `as_sign` in future when we support more signing algorithms.

iliana replied:
OK, I think I see what's necessary here. We need to support a way of saying "use this particular signing algorithm with KMS" in case the TUF spec adds additional signing algorithms for the same kind of key.

If that's the case, it needs to be a member of `KmsKeySource` so that the user can specify a different algorithm. And I would avoid using strings at all in this circumstance.

Here's what I'd propose:

```rust
#[non_exhaustive]
pub enum KmsSigningAlgorithm {
    RsassaPssSha256,
}

pub struct KmsKeySource {
    ...
    pub signing_algorithm: KmsSigningAlgorithm,
}
```

Then, verify the GetPublicKey response contains a value that matches `signing_algorithm`, returning an error if there's a mismatch (perhaps the key in KMS is ECDSA).

Because there's only one option implemented for now, I don't think we need to worry about wiring this up all the way through tuftool yet.

srgothi92 replied:
Done.

iliana commented on tough-kms/src/lib.rs:
```diff
@@ -0,0 +1,171 @@
+}
+
+/// Implements the Sign trait for KMS rsa Key
+pub struct KmsRsaKey {
+    /// Key Id of Customer Managed Key in KMS used to sign the message
+    pub key_id: String,
+    /// Aws account profile
+    pub profile: Option<String>,
+    /// KmsClient Object to query AWS KMS
+    pub client: Option<KmsClient>,
+    /// Public Key corresponding to Customer Managed Key
+    pub public_key: Decoded<RsaPem>,
+    /// Signing Algorithm to be used for the Customer Managed Key
+    pub signing_algorithm: KmsSigningAlgorithms,
+}
```
I don't think these fields should be `pub` — anything going on here is purely an implementation detail.

srgothi92 replied:
Done.

iliana commented on tough-kms/src/lib.rs:
```diff
@@ -0,0 +1,171 @@
+        };
+        let sign_fut = kms_client.sign(SignRequest {
+            key_id: self.key_id.clone(),
+            message: bytes::Bytes::from(msg.to_owned()),
+            message_type: Some(String::from("RAW")),
```
Apparently [messages can only be 4096 bytes long](https://docs.aws.amazon.com/kms/latest/APIReference/API_Sign.html#KMS-Sign-request-Message), so we will want to use digest mode here.

Along with a `use ring::digest::{digest, SHA256};` up top, I've replaced this with:

```suggestion
            message: digest(&SHA256, msg).as_ref().to_vec().into(),
            message_type: Some(String::from("DIGEST")),
```

(This doesn't fix the mysterious invalid signature issue, which I'm working through separately.)

srgothi92 replied:
Done. I wanted to use "DIGEST" but couldn't figure out `digest(&SHA256, msg).as_ref().to_vec().into()` part

iliana reviewed (commented):
A first pass of thoughts.

I'm unable to verify signatures produced by this code. I'm investigating why, because the answer is not obvious to me.

srgothi92 commented:
Fixed some of the code review comments, with multiple stupid force pushes. you might have to look at two of them to see the complete changes.

iliana committed:
tough-kms: Fix handling of public key

The DER-encoded document we are getting from the GetPublicKey API is in
the SPKI format. `Decoded<RsaPem>`'s string form is SPKI, but its byte
form is RSAPublicKey.

Previously this code created the Decoded object from the byte form,
which caused it to be double-wrapped in an SPKI document.

Because we don't have access to tough's SPKI-parsing logic (not that any
other crates should), we encode the certificate as a PEM and create the
Decoded object from the string form.

iliana commented:
ae2f466 fixes the verification issue! The commit message explains, but when we initially discussed how to go about creating the `Decoded<RsaPem>` I forgot that the byte form of it needs to be an `RSAPublicKey` document, not a `SubjectPublicKeyInfo` document. This confusion caused us to double-wrap the public key in the SPKI wrapper, resulting in this extremely confusing situation:

```
$ jq -r .signed.keys.a6c9b56dd7ce0a87fab1fa1c39d37afefcfb0692c05b2ddaa207919f4d1173d1.keyval.public ~/tmp/root.json \
> | openssl asn1parse
    0:d=0  hl=4 l= 442 cons: SEQUENCE          
    4:d=1  hl=2 l=  13 cons: SEQUENCE          
    6:d=2  hl=2 l=   9 prim: OBJECT            :rsaEncryption
   17:d=2  hl=2 l=   0 prim: NULL              
   19:d=1  hl=4 l= 423 prim: BIT STRING        
$ jq -r .signed.keys.a6c9b56dd7ce0a87fab1fa1c39d37afefcfb0692c05b2ddaa207919f4d1173d1.keyval.public ~/tmp/root.json \
> | openssl asn1parse -strparse 19
    0:d=0  hl=4 l= 418 cons: SEQUENCE          
    4:d=1  hl=2 l=  13 cons: SEQUENCE          
    6:d=2  hl=2 l=   9 prim: OBJECT            :rsaEncryption
   17:d=2  hl=2 l=   0 prim: NULL              
   19:d=1  hl=4 l= 399 prim: BIT STRING        
```

iliana commented on tough-kms/tests/data/response_public_key.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "PublicKey" : "Ii0tLS0tQkVHSU4gUFVCTElDIEtFWS0tLS0tXG5NSUlCb2pBTkJna3Foa2lHOXcwQkFRRUZBQU9DQVk4QU1JSUJpZ0tDQVlFQW5MNnU2UTlRNnBnMUc1MDIwYTgzXG5HbEgvYUZVTzBQUTVsZUlwd1dMOGtXZ3BhV3VVRzdvUmxPVUcyLzRjd041RkN2SkpHWHFVNUF0U0txMmZaNDJKXG41WFI5UU1pcDRQZzBRNm1FOFhDdkFYQW9NbmtXU2NoZHpnVDJHb0VudGFPZVJSVENVR2IvRHNWb3hzVlhqVjZtXG5GYVJNeDduaDhnZ3NoTVdnVFlnVFVESytDU0lCQ2NCV2FwQ0ZxMUJyTTYwWFptR1RxZUF1SFNIYVVVdU ...
```
This doesn't seem accurate based on the response I see from the KMS API. This is a base64'd PEM-formatted key, and the API seems to return a base64'd DER-formatted key.

srgothi92 replied:
Done

iliana commented on tuftool/tests/create_repository_integration.rs:
```diff
@@ -0,0 +1,240 @@
+#[cfg_attr(not(feature = "integration-tests"), ignore)]
+// Ensure we can use ssm key to create and sign a repo created by the `tuftool` binary using the `tough` library
+fn create_repository_ssm_key() {
+    let root_key = &format!("aws-ssm://{}/tough-integ/key-a", PROFILE);
+    create_repository(root_key, true);
```
Now that I understand what's going on here: it might be good to confirm that you can run this twice in a row. I assume tough-ssm will happily override a given parameter name.

srgothi92 replied:
Yes you can run twice in a row and `tough-ssm` will override the given parameter name.

iliana commented on tuftool/tests/create_repository_integration.rs:
```diff
@@ -0,0 +1,240 @@
+use tough::{ExpirationEnforcement, Limits, Repository, Settings};
+
+// This file include integration tests for KeySources: tough-ssm, tough-kms and local file key.
+// Since the tests are run using the actual "AWS SSM and AWS KMS", you would have to configure
+// AWS account profile "test-profile" with root permission.
+// Refer https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html to configure named profile.
```
I would prefer to use the default profile here, actually. People can run with `AWS_PROFILE=test-profile cargo test --features=integ`.

The default credential searcher should be preferred because we don't know how people develop this, and if developers want to use EC2 IMDS credentials or environment credentials, we shouldn't get in their way of testing.

iliana replied:
(Clarifying -- default credential provider, which will go through profiles, IMDS, environment variables, etc)

srgothi92 replied:
Done

iliana commented on tuftool/tests/create_repository_integration.rs:
```diff
@@ -0,0 +1,240 @@
+use tempfile::TempDir;
+use tough::{ExpirationEnforcement, Limits, Repository, Settings};
+
+// This file include integration tests for KeySources: tough-ssm, tough-kms and local file key.
```
These integration tests should also attempt to load the repositories.

srgothi92 replied:
It does, that's how I got to know verify is failing.

iliana replied:
Oh! :)

srgothi92 committed:
Fixed PR comments for and Test case for toug-kms after signature verification fix

srgothi92 commented:
PR comments fixes.

iliana commented:
> We just realized that we probably shouldn't create a tokio runtime inside these libraries, so we probably shouldn't ship this until #213 is fixed. :(

I'd rather merge this first, then do the async party. :)

iliana commented on tough-kms/src/lib.rs:
```diff
@@ -0,0 +1,203 @@
+    RsassaPssSha256,
+}
+
+impl Default for KmsSigningAlgorithm {
+    fn default() -> Self {
+        KmsSigningAlgorithm::RsassaPssSha256
+    }
+}
```
We shouldn't do this; we want the user of the key source to always be aware of what key signing algorithm is in use, and a default implementation makes less sense when we eventually add ECDSA support to tough-kms.

For now, tuftool can hardcode the `RsassaPssSha256` variant; we'll figure out a solution there for ECDSA support.

```suggestion
```

srgothi92 replied:
Done.

iliana commented on tough-kms/src/lib.rs:
```diff
@@ -0,0 +1,203 @@
+                line_ending: pem::LineEnding::LF,
+            },
+        );
+        if !response
+            .signing_algorithms
+            .context(error::MissingSignAlgorithm)?
+            .contains(&self.signing_algorithm.value())
+        {
+            error::ValidSignAlgorithm.fail()?;
+        }
```
Prefer snafu's [`ensure!`](https://docs.rs/snafu/0.6.8/snafu/macro.ensure.html) macro.

```suggestion
        ensure!(
            !response
                .signing_algorithms
                .context(error::MissingSignAlgorithm)?
                .contains(&self.signing_algorithm.value()),
            error::ValidSignAlgorithm,
        );
```

srgothi92 replied:
Done.

iliana commented on tough-kms/src/lib.rs:
```diff
@@ -0,0 +1,203 @@
+    fn value(self) -> String {
+        // Currently we are supporting only single algorithm, but code stub is added to support
+        // multiple algorithms in future.
+        match self {
+            #![allow(clippy::wildcard_in_or_patterns)]
+            #[allow(unreachable_patterns)]
+            KmsSigningAlgorithm::RsassaPssSha256 | _ => String::from("RSASSA_PSS_SHA_256"),
+        }
```
```suggestion
        String::from(match self {
            KmsSigningAlgorithm::RsassaPssSha256 => "RSASSA_PSS_SHA_256",
        })
```

srgothi92 replied:
Done.

iliana commented on tough-kms/src/lib.rs:
```diff
@@ -0,0 +1,203 @@
+}
+
+/// Implements the `KeySource` trait for keys that live in AWS KMS
+#[derive(Default)]
```
```suggestion
```

srgothi92 replied:
Done.

iliana commented on tuftool/tests/create_repository_integration.rs:
```diff
@@ -0,0 +1,246 @@
+// To run test include feature flag 'integ' like : "cargo test --features=integ"
+
+fn get_profile() -> String {
+    match env::var("AWS_PROFILE") {
+        Ok(value) => value,
+        Err(_e) => String::from("default"),
+    }
```
If we provide an empty string for the profile, it will use the default credentials provider, which permits usage of environment/IMDS credentials and the like. (And the `Default` impl for String is an empty string, so...)

```suggestion
    env::var("AWS_PROFILE").unwrap_or_default()
```

srgothi92 replied:
That was way simplified 👍 . Done.

iliana reviewed (commented):
Final stretch!

srgothi92 commented:
Fixed PR comments. Check 1st `force-pushed` for all relevant changes, 2nd was for missed `cargo fmt -- --check`

zmrow reviewed (approved):
🌃 

lgtm

iliana committed:
Merge branch 'develop' into develop
